### PR TITLE
chore 📦: End-to-end speaker diarization with reliability and config hardening

### PR DESCRIPTION
**Pull request by beecave-homelab**
# Pull Request: End-to-end speaker diarization with reliability and config hardening

## Summary

This PR merges `dev` into `main`, promoting end-to-end speaker diarization support plus follow-up reliability fixes across API, CLI, WebUI, dependency resolution, and tests. Net effect: `main` receives the diarization feature set, regression fixes from review feedback, and aligned docs/config for production usage.

## Why

- Promote completed diarization work from `dev` to `main` in one reviewable merge.
- Ship post-review fixes (cache key isolation by HF token, validation hardening, error-path cleanup).
- Keep runtime/docs/dependency behavior consistent across API, CLI, and WebUI entrypoints.

## Notable changes (reviewer focus)

- Added new diarization integration module with pipeline caching, token-aware cache keys, ROCm-safe audio-loading fallback, and speaker/chunk alignment.
- Unified post-processing behavior (stabilization + diarization) in API routes and tightened diarization device validation paths.
- Extended CLI/WebUI diarization options and defaults to align with centralized constants and auto-detect speaker behavior.
- Updated dependency/export artifacts (`pyproject.toml`, lockfile, requirements, Dockerfiles) to support diarization extras in reproducible builds.
- Added/expanded API, CLI, core integration, and WebUI tests for diarization behavior and regressions.
- Updated docs and environment examples (`README.md`, `.env.example`, `AGENTS.md`, `VERSIONS.md`, `project-overview.md`) to reflect new runtime configuration.

## Files changed

- **Counts:** 41 files changed (6 added, 35 modified, 0 deleted).
- **Key touchpoints:** `api/`, `cli/`, `core/integrations/`, `webui/`, dependency manifests, docs, and test suites.

<details>
<summary>File lists (auto)</summary>

### Added
- `insanely_fast_whisper_rocm/core/integrations/diarization.py`
- `tests/api/conftest.py`
- `tests/api/test_api_diarization.py`
- `tests/core/__init__.py`
- `tests/core/integrations/__init__.py`
- `tests/core/integrations/test_diarization.py`

### Modified
- `.coderabbit.yaml`
- `.env.example`
- `.windsurf/rules/testing.md`
- `AGENTS.md`
- `Dockerfile`
- `Dockerfile.dev`
- `README.md`
- `VERSIONS.md`
- `insanely_fast_whisper_rocm/api/models.py`
- `insanely_fast_whisper_rocm/api/responses.py`
- `insanely_fast_whisper_rocm/api/routes.py`
- `insanely_fast_whisper_rocm/cli/cli.py`
- `insanely_fast_whisper_rocm/cli/commands.py`
- `insanely_fast_whisper_rocm/cli/common_options.py`
- `insanely_fast_whisper_rocm/core/asr_backend.py`
- `insanely_fast_whisper_rocm/core/errors.py`
- `insanely_fast_whisper_rocm/core/formatters.py`
- `insanely_fast_whisper_rocm/core/integrations/__init__.py`
- `insanely_fast_whisper_rocm/core/segmentation.py`
- `insanely_fast_whisper_rocm/utils/__init__.py`
- `insanely_fast_whisper_rocm/utils/constants.py`
- `insanely_fast_whisper_rocm/webui/app.py`
- `insanely_fast_whisper_rocm/webui/handlers.py`
- `insanely_fast_whisper_rocm/webui/ui.py`
- `pdm.lock`
- `project-overview.md`
- `pyproject.toml`
- `requirements-all.txt`
- `requirements-rocm-v7-0.txt`
- `scripts/hf_models.py`
- `tests/api/test_models.py`
- `tests/cli/test_cli.py`
- `tests/utils/test_centralized_configuration.py`
- `tests/webui/test_app.py`
- `tests/webui/test_ui.py`

### Deleted
- `N/A`

</details>

## Test plan

- [x] Unit: Added/updated diarization unit tests in core + CLI argument/validation coverage.
- [x] Integration: Added API and WebUI diarization integration tests; exercised diarization error/success paths.
- [ ] Manual: N/A for this merge PR (validated via automated tests and prior PR verification).

## Risks & rollback

- **Risks:**
  - Dependency/runtime matrix changes (diarization extras + lockfile/requirements updates) may affect build reproducibility if environment drift exists.
  - New diarization failure behavior now surfaces explicit errors when dependencies are missing.
  - Merge includes broad cross-layer changes (API/CLI/WebUI), so regressions may surface in less-traveled paths.
- **Rollback:** Revert this merge commit on `main` and redeploy the previous stable image/tag.

## Additional notes

- This is a promotion PR from `dev` after PR #58 merge and follow-up fixes.
- If desired, I can run a draft `gh pr create` dry-run next using this body file.

## Summary by CodeRabbit

* **New Features**
  * Added optional speaker diarization to transcription workflows, with controls for speaker counts and processing device.
  * Speaker labels and diarization status now appear in JSON, TXT, SRT, and VTT outputs, including translations.
* **Improvements**
  * Added timestamp-stabilization status and persistent post-processing results.
  * Improved progress updates, error reporting, and GPU/CPU processing resilience.
  * Expanded Web UI result displays and JSON downloads. The save directory is now read-only and administrator-configured.
  * Improved punctuation spacing in subtitle and text outputs.
* **Documentation**
  * Updated CLI branding and help text.

### CODE REVIEW

**Review by beecave-orchestrator — APPROVED** [review]
## Review: PR #59 — Diarization promotion (dev → main)

**Verdict: APPROVE**

This is a well-structured feature promotion with solid test coverage and proper isolation of the optional pyannote.audio dependency. CI is green across all checks.

### Positive

- Clean separation: diarization is fully optional with graceful degradation
- Good pyannote v3/v4 compatibility (`hasattr` check for `speaker_diarization`)
- ROCm torchcodec workaround is defensive (catches both `ImportError` and `OSError`)
- Lazy import in API routes avoids cold-start penalty when diarization is not used
- Solid test coverage: core logic, API integration, CLI parsing, caching, alignment
- HF_TOKEN fallback chain (`HF_TOKEN` → `HUGGINGFACE_TOKEN` → `HUGGINGFACE_HUB_TOKEN`)
- Dockerfile `--no-deps` correctly prevents torchcodec re-introduction

### Should-fix (non-blocking, follow-up OK)

1. **Cache key excludes `hf_token`.** `_get_or_create_pipeline` caches on `(model_name, device)` but ignores the token. Changing `HF_TOKEN` between calls reuses the old pipeline. Add a token hash to the key.

2. **Duplicated diarization blocks in routes.** `create_transcription` and `create_translation` contain near-identical 15-line diarization blocks. Extract a shared helper to prevent divergence.

3. **`diarization_device` not validated at API boundary.** `VALID_DIARIZATION_DEVICES` is defined but never used in validation. A malformed device string will fail deep in pyannote with an unclear error.

4. **WebUI device radio hardcodes `["cpu", "cuda"]`.** Works on ROCm (PyTorch uses `"cuda"` for AMD GPUs) but could confuse users. A label or tooltip would help.

### Nice-to-have

- `_majority_speaker` re-imports `Counter` on every call — module-level import would be cleaner
- Tests duplicate `torchaudio.load` mock setup instead of sharing a fixture
- Missing test for `_align_speakers_to_segments` with `timestamp=(None, None)` chunks

### Dependency note

- **protobuf downgrade:** 7.34.0 → 6.33.6. Verify this does not break onnxruntime-rocm.
- Heavy pyannote dependency tree is properly isolated in the `diarization` PDM group.

---

Verdict card: `agent-output/code-reviewer/verdict-cards/pr59-diarization-promotion.md`

**Review by beecave-orchestrator — CHANGES_REQUESTED** [review]
## Review: PR #59 — Diarization promotion (dev → main) — **Updated after live testing**

**Previous review (4177410611) approved this PR. This update revises that verdict based on end-to-end testing on a live deployment.**

**Verdict: REQUEST_CHANGES**

### What changed since the original review

I tested the `dev` branch on `192.168.2.110:7862` with a ~28 min multi-speaker Dutch audio file (`PooleTeams.m4a`). Diarization was enabled via the WebUI (min 1, max 4 speakers, cuda device).

**The SRT output contains no `[SPEAKER_XX]:` prefixes.** The produced file (471 segments) is identical to a non-diarized run. The server logs confirm:

```
[SrtFormatter] Returning 471 SRT segments.
[_add_files_flat] Formatted content for Poolteams.srt (len 35301): '1
00:00:00,224 --> 00:00:02,812
...werder ook geen
verdeling meer in plaats.'
```

The `result_data` includes `diarization_error` and `diarized` keys, so the diarization step ran but either failed (setting `diarized=False`) or the chunks reaching the formatters lack `speaker` fields.

### Must-fix

1. **Diarization failure is invisible to users.** When `diarize_result()` catches an error in `handlers.py`, it sets `result["diarized"] = False` and logs a warning. The WebUI transcription summary, SRT, TXT, and JSON outputs give zero indication that diarization was requested but failed. The user sees the same output as if they never enabled it. Surface the `diarization_error` in the WebUI summary text at minimum.

2. **Verify diarized chunks propagate to zip creator.** The `diarize()` integration in `core/integrations/diarization.py` annotates chunks with `speaker` fields. The `zip_creator` calls formatters with the same `result_data` dict. But the formatter guards (`if segment.speaker:`, `if speaker:`) silently skip when speaker is `None`, producing output identical to non-diarized runs. Either the diarization step is failing before annotating chunks, or the annotated data isn't surviving the path from `diarize_result()` → `result` → `zip_creator` → formatter. Trace this path and fix or add logging.

### Previously flagged (still valid, non-blocking)

3. **Cache key excludes `hf_token`** (from original review)
4. **Duplicated diarization blocks in API routes** (from original review)
5. **`diarization_device` not validated at API boundary** (from original review)

### Previously positive (still true)

- Clean separation of optional pyannote dependency
- Lazy imports avoid cold-start penalty
- Good test coverage for core diarization logic
- HF_TOKEN fallback chain is sensible

---

Verdict card: `agent-output/code-reviewer/verdict-cards/pr59-diarization-promotion-v2.md`

**Review by beecave-orchestrator — APPROVED** [review]
Changes fixed. Approved.

**Comment by bestonephoenix** [comment]
Hello, sorry for the long time to reply. I just rebuilt the image and tested the path:

I am getting 
ROCM_PATH=/opt/rocm
rocRAND header: MISSING

Should I share my compose, as I am pretty sure I am mounting the rocm devices properly.

> Thanks for testing this and posting the logs @bestonephoenix.
> 
> I dug into the error a bit more. And it looks like this is not really an ONNX problem. MIOpen is trying to compile a GPU kernel at runtime and cannot find one of the ROCm rocRAND headers:text
> 
> ```
> rocrand/rocrand_xorwow.h file not found
> ```
> 
> That points at the ROCm/MIOpen stack rather than the diarization code itself. There is an upstream MIOpen issue for the same missing-include-path problem: [ROCm/MIOpen#3314](https://github.com/ROCm/MIOpen/issues/3314)
> 
> The CPU retry is expected here. It is why the transcription still finishes instead of failing completely.
> 
> One thing I would like to rule out first: the Docker image should set `ROCM_PATH=/opt/rocm`, but your logs suggest the ROCm runtime inside the container may not be seeing everything it expects. Could you run this against the running app container and paste the output?
> 
> ```shell
> docker compose exec insanely-fast-whisper-rocm sh -lc '
>   echo "ROCM_PATH=$ROCM_PATH"
>   test -f /opt/rocm/include/rocrand/rocrand_xorwow.h \
>     && echo "rocRAND header: present" \
>     || echo "rocRAND header: MISSING"
> '
> ```
> 
> If the variable or header is missing, we can focus on the image, container, or Proxmox LXC setup. If both are present, the likely issue is the installed ROCm/MIOpen version and its runtime compiler path on gfx1100.
> 
> So I do not think this means there is anything wrong with your GPU. I can help make the project-side checks and guidance clearer, but the actual compile failure sits in the AMD ROCm/MIOpen layer.

File: tests/api/test_api_diarization.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                {
+                    "start": 0.0,
+                    "end": 2.0,
+                    "text": "Hello world.",
+                    "speaker": "SPEAKER_00",
+                },
+            ],
+            "diarized": True,
+        },
+    ):
```
[the remark this answers is not in the document]
Comment (beecave-homelab) [reply tests/api/test_api_diarization.py]: The patch target is correct as written — no change needed here.

`routes.py` imports `diarize as diarize_result` **inside** `_apply_post_processing` (see `insanely_fast_whisper_rocm/api/routes.py:166-168`), so the name is re-resolved from the `insanely_fast_whisper_rocm.core.integrations.diarization` module attribute on every call. That is exactly the symbol the test patches (`insanely_fast_whisper_rocm.core.integrations.diarization.diarize`), so the mock takes effect on each request.

Verification: `pdm run pytest tests/api/test_api_diarization.py -q` → **13 passed**, including `test_post_transcriptions__maps_diarization_error_to_400` and `test_post_transcriptions__records_diarization_error_in_saved_json`, which assert HTTP 400 and the mocked `DiarizationError` message — those assertions only pass if the patched callable is the one the route actually invokes.

If the import were hoisted to module scope your point would hold, but because it is function-local the current patch target is the right one. Marking this as a false positive.

File: insanely_fast_whisper_rocm/core/integrations/diarization.py [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+        raise DiarizationError(
+            f"Diarization inference failed: {exc}",
+            model=DEFAULT_DIARIZATION_MODEL,
+            reason="inference_error",
+        ) from exc
+
+    finally:
+        # Release VRAM even when inference or the CPU retry fails.
+        if device == "cuda":
+            _clear_gpu_diarization_cache()
```
[the remark this answers is not in the document]
Comment (beecave-homelab) [reply insanely_fast_whisper_rocm/core/integrations/diarization.py]: Thanks—confirmed this race was possible because the cache lock did not cover pipeline use. Commit 2100f96 adds per-pipeline active-use tracking and defers GPU migration until the final active inference releases the shared pipeline, including error and CPU-fallback paths. A regression test covers overlapping users and verifies migration happens exactly once after the final release. Validation: Ruff passed; 39 diarization tests and 297 core tests passed.

File: insanely_fast_whisper_rocm/core/formatters.py [hunk 3]
```diff
@@ -242,10 +269,26 @@ def get_file_extension(cls) -> str:
...
+    def _join_fragments(fragments: list[str]) -> str:
+        """Join word or segment fragments into readable prose.
+
+        Args:
+            fragments: Adjacent text fragments from one speaker turn.
+
+        Returns:
+            A normalized paragraph without spaces before punctuation.
+        """
+        text = "".join(fragments).strip()
```
[the remark this answers is not in the document]
Comment (beecave-homelab) [reply insanely_fast_whisper_rocm/core/formatters.py]: Fixed in 8f4be5a. Fragment joining now inserts a separator only when adjacent whitespace-free alphanumeric fragments would merge, while adjacent CJK-script fragments still concatenate directly. Added regressions for both `Hello` + `world` and `你` + `好`; 73 formatter tests and repository-wide Ruff checks pass.
